examples/fabric-admin/device_manager/DeviceSubscriptionManager.cpp

```diff
@@ -34,22 +34,53 @@ namespace {
 
 void OnDeviceConnectedWrapper(void * context, Messaging::ExchangeManager & exchangeMgr, const SessionHandle & sessionHandle)
 {
-    reinterpret_cast<DeviceSubscription *>(context)->OnDeviceConnected(exchangeMgr, sessionHandle);
+    reinterpret_cast<DeviceSubscriptionManager::DeviceSubscription *>(context)->OnDeviceConnected(exchangeMgr, sessionHandle);
 }
 
 void OnDeviceConnectionFailureWrapper(void * context, const ScopedNodeId & peerId, CHIP_ERROR error)
 {
-    reinterpret_cast<DeviceSubscription *>(context)->OnDeviceConnectionFailure(peerId, error);
+    reinterpret_cast<DeviceSubscriptionManager::DeviceSubscription *>(context)->OnDeviceConnectionFailure(peerId, error);
 }
 
 } // namespace
 
-DeviceSubscription::DeviceSubscription() :
+DeviceSubscriptionManager & DeviceSubscriptionManager::Instance()
+{
+    static DeviceSubscriptionManager instance;
+    return instance;
+}
+
+CHIP_ERROR DeviceSubscriptionManager::StartSubscription(Controller::DeviceController & controller, NodeId nodeId)
+{
+    auto it = mDeviceSubscriptionMap.find(nodeId);
+    VerifyOrReturnError((it == mDeviceSubscriptionMap.end()), CHIP_ERROR_INCORRECT_STATE);
+
+    auto deviceSubscription = std::make_unique<DeviceSubscription>();
+    VerifyOrReturnError(deviceSubscription, CHIP_ERROR_NO_MEMORY);
+    ReturnErrorOnFailure(deviceSubscription->StartSubscription(this, controller, nodeId));
+
+    mDeviceSubscriptionMap[nodeId] = std::move(deviceSubscription);
+    return CHIP_NO_ERROR;
+}
+
+void DeviceSubscriptionManager::DeviceSubscriptionTerminated(NodeId nodeId)
+{
+    auto it = mDeviceSubscriptionMap.find(nodeId);
+    // DeviceSubscriptionTerminated is a private method that is expected to only
```

Comment:
CHIP_ERROR RemoveSubscription(chip::NodeId nodeId) {
        std::lock_guard<std::mutex> lock(mMapMutex);
        auto it = mDeviceSubscriptionMap.find(nodeId);
        if (it != mDeviceSubscriptionMap.end()) {
            // Assuming we add a method to cleanup the subscription
            it->second->Cleanup();
            mDeviceSubscriptionMap.erase(it);
            return CHIP_NO_ERROR;
        }
        return CHIP_ERROR_NOT_FOUND;
    }